Disallow to execute cheat commands on server with disabled cheat commands

BannZay opened the pull request:
Possible improvement here:
Add a possible to read value marked by CMDFLAG_TEST. I can not find not ungly solution to make it happen
related to #2773

BannZay committed:
Write to console if rcon command was blocked

BannZay committed:
Introduce CFGFLAG_READONLY

BannZay committed:
allow map to change test settings

BannZay committed:
add CMDFLAG_TEST to registrations

heinrich5991 commented on src/engine/shared/config_variables.h:
```diff
@@ -246,41 +246,40 @@ MACRO_CONFIG_STR(SvRulesLine8, sv_rules_line8, 128, "", CFGFLAG_SERVER, "Rules l
-MACRO_CONFIG_INT(SvTeamLock, sv_team_lock, 1, 0, 1, CFGFLAG_SERVER|CFGFLAG_GAME, "Enable team lock")
+MACRO_CONFIG_INT(SvTeam, sv_team, 1, 0, 3, CFGFLAG_SERVER|CFGFLAG_GAME|CMDFLAG_TEST, "Teams configuration (0 = off, 1 = on but optional, 2 = must play only with teams, 3 = forced random team only for you)")
+MACRO_CONFIG_INT(SvTeamMaxSize, sv_max_team_size, MAX_CLIENTS, 1, MAX_CLIENTS, CFGFLAG_SERVER|CFGFLAG_GAME|CMDFLAG_TEST, "Maximum team size (from 2 to 64)")
+MACRO_CONFIG_INT(SvTeamLock, sv_team_lock, 1, 0, 1, CFGFLAG_SERVER, "Enable team lock")
```
Needs going through maps and check whether some map contains this config.

BannZay replied:
no map contain this setting

BannZay replied:
Done. 
Any CFGFLAG_GAME is considered CFGFLAG_TEST now. CFGFLAG_TEST removed.
To prevent modification by the map config you can use CFGFLAG_NOMAPCFG now.

heinrich5991 commented on src/engine/shared/config_variables.h:
```diff
@@ -178,16 +178,16 @@ MACRO_CONFIG_STR(DbgStressServer, dbg_stress_server, 32, "localhost", CFGFLAG_CL
 MACRO_CONFIG_INT(SvHideScore, sv_hide_score, 0, 0, 1, CFGFLAG_SERVER, "Whether players scores will be announced or not")
-MACRO_CONFIG_INT(SvSaveWorseScores, sv_save_worse_scores, 1, 0, 1, CFGFLAG_SERVER|CFGFLAG_GAME, "Whether to save worse scores when you already have a better one")
-MACRO_CONFIG_INT(SvPauseable, sv_pauseable, 0, 0, 1, CFGFLAG_SERVER|CFGFLAG_GAME, "Whether players can pause their char or not")
+MACRO_CONFIG_INT(SvSaveWorseScores, sv_save_worse_scores, 1, 0, 1, CFGFLAG_SERVER, "Whether to save worse scores when you already have a better one")
```
Also needs going through maps and check whether some map contains this config.

BannZay replied:
it used in a single map called Flappy Bird

BannZay replied:
so we do let this setting be available for editing by the map or we can edit "Flappy Bird"?

heinrich5991 replied:
Flappy Bird is … special. I don't know about the original reasoning here. Maybe @def- knows, but he's on holidays right now. Maybe that setting can simply be dropped, polluting the DB a little.

BannZay replied:
I make it possible to change read only variables by the map. Not sure how good decision it is but it solves this problem.
If Flappy bird can be modified we might remove the ability to change readonly variables by the map later

BannZay replied:
I added flag back. But should the map be able to control this setting is a good question :)

def- replied:
For Flappy Bird we needed it because otherwise someone can make way too many scores and we would end up with a huge database. I think it's worth having this for maps that are super-easy to finish.

heinrich5991 commented on src/engine/shared/config_variables.h:
```diff
@@ -318,17 +317,16 @@ MACRO_CONFIG_INT(SvNamelessScore, sv_nameless_score, 1, 0, 1, CFGFLAG_SERVER, "W
 MACRO_CONFIG_INT(SvReloadWhenEmpty, sv_reload_when_empty, 0, 0, 2, CFGFLAG_SERVER, "Reload map when server is empty (1 = reload once, 2 = reload every time server gets empty)")
 MACRO_CONFIG_INT(SvKillProtection, sv_kill_protection, 20, 0, 9999, CFGFLAG_SERVER, "0 - Disable, 1-9999 minutes")
-MACRO_CONFIG_INT(SvSoloServer, sv_solo_server, 0, 0, 1, CFGFLAG_SERVER|CFGFLAG_GAME, "Set server to solo mode (no player interactions, has to be set before loading the map)")
+MACRO_CONFIG_INT(SvSoloServer, sv_solo_server, 0, 0, 1, CFGFLAG_SERVER|CFGFLAG_GAME|CFGFLAG_READONLY, "Set server to solo mode (no player interactions, has to be set before loading the map)")
```
Is this set in maps?

BannZay replied:
yes in 40/311

BannZay replied:
I make it possible to change read only variables by the map. Not sure how good decision it is but it solves this problem.
If Flappy bird can be modified we might remove the ability to change readonly variables by the map later

BannZay replied:
I make it possible to change read only variables by the map. Not sure how good decision it is but it solves this problem

heinrich5991 commented on src/engine/shared/console.cpp:
```diff
@@ -736,6 +740,12 @@ static void IntVariableCommand(IConsole::IResult *pResult, void *pUserData)
 
 	if(pResult->NumArguments())
 	{
+		if (pData->m_Readonly)
```
```suggestion
		if(pData->m_Readonly)
```

heinrich5991 commented on src/engine/shared/console.cpp:
```diff
@@ -506,7 +509,7 @@ void CConsole::ExecuteLineStroked(int Stroke, const char *pStr, int ClientID, bo
 						}
 
-						if (pCommand->m_Flags&CMDFLAG_TEST)
+						if (ClientID != CLIENT_ID_GAME && pCommand->m_Flags&CMDFLAG_TEST)
```
```suggestion
						if(ClientID != CLIENT_ID_GAME && pCommand->m_Flags&CMDFLAG_TEST)
```

BannZay replied:
can we run replace all on "if (" to "if(" on sources because there is many cases I see

BannZay replied:
Resolved btw

heinrich5991 reviewed (commented):
It seems that all `CFGFLAG_GAME` commands were also marked as `CFGFLAG_TEST`. Since you probably want to maintain that going forward, it might be better to not mark them as `CFGFLAG_TEST` but rather disallow modification of `CFGFLAG_GAME` variables by players unless `sv_test_cmds` is enabled.

BannZay committed:
introduce CFGFLAG_NOMAPCFG

BannZay committed:
fix formatting

BannZay committed:
mark sv_save_worse_scores by CFGFLAG_GAME

BannZay committed:
Merge branch 'master' into NotifyAboutCommandWasRefusedToExecute

fokkonaut commented:
Lol, you add CFGFLAG_GAME and CFGFLAG_NOMAPCFG together? That doesnt make sense. CFGFLAG_GAME is to make a setting work from inside of maps.

BannZay commented:
> CFGFLAG_GAME is to make a setting work from inside of maps

Are you sure ? Why does this flag have such name then?

> you add CFGFLAG_GAME and CFGFLAG_NOMAPCFG together?

@fokkonaut , PR name is not relevant anymore, true. You could check commit history and comments to them, but I will explain why I did what I did.

What was done:
1. Added CMDFLAG_TEST to all occurences CFGFLAG_GAME to prevent using commands like "tune" in non tests server mode.
2.  > it might be better to not mark them as CFGFLAG_TEST but rather disallow modification of CFGFLAG_GAME
3. CMDFLAG_TEST become useless, but we stil wanted to forbid some commands from changing in map cfg so I added CFGFLAG_NOMAPCFG.

> BannZay09/09/2020
The reason behind: to me CFGFLAG_GAME means "this command affects gameplay". So we should mark commands like "tele, shotgun" by CFGFLAG_GAME
we might introduce CFGFLAG_NOMAPCFG or something (if ever need). But with CFGFLAG_TEST I have no idea how to understand these flags and how to mark without marking a lot of commadns by both of em
heinrich599109/09/2020
hm. might be possible, at the expense of less meaningful error messages
one could just leave the flags as either GAME or TEST though
that would make it easy to see which to apply

fokkonaut commented:
I handled it like this in my mod: https://github.com/fokkonaut/F-DDrace/blob/b50677f4d62e192d22937573fe3555a3e4ee41fc/src/engine/shared/console.cpp#L422

BannZay committed:
fix CFGFLAG_NOMAPCFG not used in check

BannZay committed:
Merge branch 'master' into NotifyAboutCommandWasRefusedToExecute

BannZay commented:
> CFGFLAG_GAME means it can be used from the game, so inside of the map.

So you say that The Game == The Map ? This is weird. I guess there is either wrong flag name or wrong flag usage or both...

@fokkonaut could you provide alternative solution to problems solved by this PR?
Why can not we treat CFGFLAG_GAME, _SERVER, _CLIENT as "command affects _GAME, _SERVER, _CLIENT" ?

fokkonaut commented:
no idea what you mean, but just read the code or try it out yourself. cfgflag_game is the map settings, yes.

> Why can not we treat CFGFLAG_GAME, _SERVER, _CLIENT as "command affects _GAME, _SERVER, _CLIENT" ?

?

BannZay commented:
> I handled it like this in my mod: https://github.com/fokkonaut/F-DDrace/blob/b50677f4d62e192d22937573fe3555a3e4ee41fc/src/engine/shared/console.cpp#L422

@fokkonaut cool. Have you solved a problem of using "sv_team" command with disabled test commands ?

def- commented:
@heinrich5991 is this change still wanted?

If so, it has some conflicts now with my pure-removal and clang-format change

fokkonaut commented:
This doesnt make any sense... A readonly flag would make sense, not this mapcfg flag. There is cfgflag_game.

heinrich5991 commented:
@fokkonaut Where's the problem in the new flag? I like the new distinction. `GAME` now means it has influence over the gameplay, and `NOMAPCFG` means that it cannot be set from within maps.

I guess the best option to reduce the amount of flags would be to move from `GAME | NOMAPCFG` back to `TEST`, commands that can only be executed when sv_test_cmds is on.

@BannZay Still interested in this?

BannZay commented:
> to reduce the amount of flags would be to move from GAME | NOMAPCFG back to TEST

Ehm... might be better to define `TEST = GAME | NOMAPCFG` and do replacements but then I guess it will be harder to understand the code. 
I prefer to explicitly have a GAME flag here. To me 'NOMAPCFG'/'Test'/whatever is secondary but `GAME` is main thing here.

> @BannZay Still interested in this?

Yeah, I can bring TEST back if you sure it is the right direction.

fokkonaut commented:
Why is this used? Can we just rename GAME to MAP? Then test stays test and its fine? The new bahviour doesnt seem useful

BannZay commented:
> Why is this used?

Current situation:
CONSOLE_COMMAND("super", "", CFGFLAG_SERVER | CMDFLAG_TEST...
MACRO_CONFIG_INT(SvTeam, sv_team, 1, 0, 3, CFGFLAG_SERVER | CFGFLAG_GAME ...

Which means it is possible to use sv_team even though test_cmds disabled. This is incorrect behavior to test_cmds.
I researched and found that each GAME or TEST command must be forbidden by test_cmds. Eventually, I marked all TEST commands by GAME as well to let test_cmds to forbid them. Later I renamed CMDFLAG_TEST to NOMAPCFG (a much more meaningful name) for code readability purposes.

I didn't want to make things overcomplicated so I made more noise in flags instead of not bvious code behind just to make things easier to understand.

@fokkonaut  I don't mind if this PR will be rejected but I am bored writing explanations to your blurry complaints.

BannZay committed:
Merge branch 'master' into NotifyAboutCommandWasRefusedToExecute

heinrich5991 commented:
Superseded by #3066.
